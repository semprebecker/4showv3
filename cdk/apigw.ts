import { WebSocketApi, WebSocketStage } from "@aws-cdk/aws-apigatewayv2-alpha";
import { WebSocketLambdaIntegration } from "@aws-cdk/aws-apigatewayv2-integrations-alpha";
import { RemovalPolicy, Stack } from "aws-cdk-lib";
import { CfnAccount } from "aws-cdk-lib/aws-apigateway";
import { CfnModel, CfnRoute, CfnStage } from "aws-cdk-lib/aws-apigatewayv2";
import {
  ManagedPolicy,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam";
import { Function as LambdaFunction } from "aws-cdk-lib/aws-lambda";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import * as cognito from "aws-cdk-lib/aws-cognito";

import { MessageAction } from "../backend/types";
import { getCfnAuthorizer } from "./authorizer";
import { lambdaFunctions } from "./lambda";
import { getModels } from "./models";
import { envs } from "../backend/main/config";

export const stageName = envs.stage;

/*
 * Utility function for adding a route with a model.
 * This is not supported in the L2 constuct so we must drop to
 * L1 (Cfn) by casting the `WebSocketRoute` to `CfnRoute` (`AWS::ApiGatewayV2::Route`)
 */
const addRoute = (
  handler: LambdaFunction,
  action: MessageAction,
  model: CfnModel,
  webSocketApi: WebSocketApi
) => {
  const route = webSocketApi.addRoute(action, {
    integration: new WebSocketLambdaIntegration(
      `${action}Integration`,
      handler
    ),
  });
  const rt = route.node.defaultChild as CfnRoute;

  rt.modelSelectionExpression = "$request.body.action";
  rt.requestModels = { [action]: model.name };

  rt.addDependsOn(model);
};

/*
 * Again, we need to extend the `WebSocketStage` construct by casting it to `CfnStage` (`AWS::ApiGatewayV2::Stage`).
 */
const getStageAndLogs = (
  scope: Stack,
  webSocketApi: WebSocketApi
): WebSocketStage => {
  // Unlike RestApi, you don't get a default `prod` stage automatically.
  const stage = new WebSocketStage(scope, `${envs.stage.toUpperCase()}Stage`, {
    autoDeploy: true,
    stageName,
    webSocketApi,
  });
  // Manage the log group
  new LogGroup(scope, "ExecutionLogs", {
    logGroupName: `/aws/apigateway/${webSocketApi.apiId}/${stageName}`,
    removalPolicy: RemovalPolicy.DESTROY,
    retention: RetentionDays.ONE_WEEK,
  });
  const log = new LogGroup(scope, "AccessLogs", {
    removalPolicy: RemovalPolicy.DESTROY,
    retention: RetentionDays.ONE_WEEK,
  });
  const cfnStage = stage.node.defaultChild as CfnStage;
  cfnStage.accessLogSettings = {
    destinationArn: log.logGroupArn,
    format: `$context.identity.sourceIp - - [$context.requestTime] "$context.httpMethod $context.routeKey $context.protocol" $context.status $context.responseLength $context.requestId`,
  };
  cfnStage.defaultRouteSettings = {
    dataTraceEnabled: true,
    loggingLevel: "INFO",
    throttlingBurstLimit: 500,
    throttlingRateLimit: 1000,
  };

  /*
   * This role is automatically created by the RestApi construct but not by WebSocketApi.
   * CfnAccount isn't even available in the `aws-cdk-lib/aws-apigatewayv2` lib so we must import `aws-cdk-lib/aws-apigateway`
   * to create the CloudWatch role.
   */
  const cwRole = new Role(scope, "CWRole", {
    assumedBy: new ServicePrincipal("apigateway.amazonaws.com"),
    managedPolicies: [
      ManagedPolicy.fromAwsManagedPolicyName(
        "service-role/AmazonAPIGatewayPushToCloudWatchLogs"
      ),
    ],
  });

  const account = new CfnAccount(scope, "Account", {
    cloudWatchRoleArn: cwRole.roleArn,
  });

  webSocketApi.node.addDependency(account);
  return stage;
};

export const getApiGateway = (
  scope: Stack,
  fns: lambdaFunctions
): [WebSocketApi, WebSocketStage] => {
  const webSocketApi = new WebSocketApi(
    scope,
    `WebSocketApi4show${envs.stage.toUpperCase()}`,
    {
      connectRouteOptions: {
        integration: new WebSocketLambdaIntegration(
          "OnConnectIntegration",
          fns["setup-connection/onConnect"]
        ),
      },
      disconnectRouteOptions: {
        integration: new WebSocketLambdaIntegration(
          "OnDisconnectIntegration",
          fns["setup-connection/onDisconnect"]
        ),
      },
    }
  );

  const {
    getStateModel,
    addUser,
    updateUser,
    getUser,
    addAddressUser,
    updateAddressUser,
    getAddressesUser,
    listProfilesAdmin,
    listPermissionsAdmin,
    manageProfilePermissionsAdmin,
    listProfilePermissionsAdmin,
    addAccountAdmin,
    updateAccountAdmin,
    deleteAccountAdmin,
    listAccountsAdmin,
    countPagesAccountsAdmin,
    manageAccountLinkedUsersAdmin,
    listAccountLinkedUsersAdmin,
    countPagesUsersAdmin,
  } = getModels(scope, webSocketApi);

  getCfnAuthorizer(scope, webSocketApi, fns["setup-connection/authorizer"]);

  addRoute(
    fns["setup-connection/getState"],
    MessageAction.GET_STATE,
    getStateModel,
    webSocketApi
  );

  addRoute(fns["user/addUser"], MessageAction.ADD_USER, addUser, webSocketApi);

  addRoute(
    fns["user/updateUser"],
    MessageAction.UPDATE_USER,
    updateUser,
    webSocketApi
  );

  addRoute(fns["user/getUser"], MessageAction.GET_USER, getUser, webSocketApi);

  addRoute(
    fns["user/addAddress"],
    MessageAction.ADD_ADDRESS_USER,
    addAddressUser,
    webSocketApi
  );

  addRoute(
    fns["user/updateAddress"],
    MessageAction.UPDATE_ADDRESS_USER,
    updateAddressUser,
    webSocketApi
  );

  addRoute(
    fns["user/getAddresses"],
    MessageAction.GET_ADDRESSES_USER,
    getAddressesUser,
    webSocketApi
  );

  addRoute(
    fns["admin/listProfiles"],
    MessageAction.LIST_PROFILES,
    listProfilesAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/listPermissions"],
    MessageAction.LIST_PERMISSIONS,
    listPermissionsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/manageProfilePermissions"],
    MessageAction.MANAGE_PROFILE_PERMISSIONS,
    manageProfilePermissionsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/listProfilePermissions"],
    MessageAction.LIST_PROFILE_PERMISSIONS,
    listProfilePermissionsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/addAccount"],
    MessageAction.ADD_ACCOUNT,
    addAccountAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/updateAccount"],
    MessageAction.UPDATE_ACCOUNT,
    updateAccountAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/deleteAccount"],
    MessageAction.DELETE_ACCOUNT,
    deleteAccountAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/listAccounts"],
    MessageAction.LIST_ACCOUNTS,
    listAccountsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/countPagesAccounts"],
    MessageAction.COUNT_PAGES_ACCOUNTS,
    countPagesAccountsAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/manageAccountLinkedUsers"],
    MessageAction.MANAGE_ACCOUNT_LINKED_USERS,
    manageAccountLinkedUsersAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/listAccountLinkedUsers"],
    MessageAction.LIST_ACCOUNT_LINKED_USERS,
    listAccountLinkedUsersAdmin,
    webSocketApi
  );

  addRoute(
    fns["admin/countPagesUsers"],
    MessageAction.COUNT_PAGES_USERS,
    countPagesUsersAdmin,
    webSocketApi
  );

  // These functions need to be able to manage websocket connections.
  [
    fns["setup-connection/getState"],
    fns["user/addUser"],
    fns["user/updateUser"],
    fns["user/getUser"],
    fns["user/addAddress"],
    fns["user/updateAddress"],
    fns["user/getAddresses"],
    fns["admin/listProfiles"],
    fns["admin/listPermissions"],
    fns["admin/manageProfilePermissions"],
    fns["admin/listProfilePermissions"],
    fns["admin/addAccount"],
    fns["admin/updateAccount"],
    fns["admin/deleteAccount"],
    fns["admin/listAccounts"],
    fns["admin/countPagesAccounts"],
    fns["admin/manageAccountLinkedUsers"],
    fns["admin/listAccountLinkedUsers"],
    fns["admin/countPagesUsers"],
  ].forEach((fn) =>
    fn.addToRolePolicy(
      new PolicyStatement({
        actions: ["execute-api:ManageConnections"],
        resources: [
          `arn:aws:execute-api:${process.env.CDK_DEFAULT_REGION}:${process.env.CDK_DEFAULT_ACCOUNT}:${webSocketApi.apiId}/*`,
        ],
      })
    )
  );

  return [webSocketApi, getStageAndLogs(scope, webSocketApi)];
};
