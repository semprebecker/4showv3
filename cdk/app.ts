import { App } from 'aws-cdk-lib';

import { SetupStack } from './setup-stack';
import { envs } from '../backend/main/config';

const app = new App();

new SetupStack(app, `4showStack${envs.stage.toUpperCase()}`, {
  description: `4show Stack ${envs.stage.toUpperCase()}`,
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  },
  stackName: `app-4show-stack-${envs.stage}`,
});
