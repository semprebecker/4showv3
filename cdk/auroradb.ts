import { Stack, Duration } from "aws-cdk-lib";
import { BastionHostLinux, IVpc, Port } from "aws-cdk-lib/aws-ec2";
import {
  Credentials,
  DatabaseClusterEngine,
  ParameterGroup,
  ServerlessCluster,
} from "aws-cdk-lib/aws-rds";
import { lambdaFunctions } from "./lambda";

import { envs } from "../backend/main/config";

export const getDatabaseService = (
  scope: Stack,
  vpc: IVpc,
  host: BastionHostLinux,
  fns: lambdaFunctions
): ServerlessCluster => {
  const cluster = new ServerlessCluster(scope, `Aurora-Serverless`, {
    engine: DatabaseClusterEngine.AURORA_POSTGRESQL,
    parameterGroup: ParameterGroup.fromParameterGroupName(
      scope,
      "ParameterGroup",
      "default.aurora-postgresql10"
    ),
    defaultDatabaseName: `${envs.stage}`,
    vpc: vpc,
    enableDataApi: true,
    scaling: { autoPause: Duration.minutes(0) },
    credentials: Credentials.fromGeneratedSecret("serverless"),
  });
  cluster.connections.allowFrom(
    host.connections,
    Port.tcp(cluster.clusterEndpoint.port),
    "Bastion host connection"
  );
  cluster.grantDataApiAccess(fns["setup-connection/getState"]);
  cluster.grantDataApiAccess(fns["user/addUser"]);
  cluster.grantDataApiAccess(fns["user/updateUser"]);
  cluster.grantDataApiAccess(fns["user/getUser"]);
  cluster.grantDataApiAccess(fns["user/addAddress"]);
  cluster.grantDataApiAccess(fns["user/updateAddress"]);
  cluster.grantDataApiAccess(fns["user/getAddresses"]);
  cluster.grantDataApiAccess(fns["admin/listProfiles"]);
  cluster.grantDataApiAccess(fns["admin/listPermissions"]);
  cluster.grantDataApiAccess(fns["admin/manageProfilePermissions"]);
  cluster.grantDataApiAccess(fns["admin/listProfilePermissions"]);
  cluster.grantDataApiAccess(fns["admin/addAccount"]);
  cluster.grantDataApiAccess(fns["admin/updateAccount"]);
  cluster.grantDataApiAccess(fns["admin/deleteAccount"]);
  cluster.grantDataApiAccess(fns["admin/listAccounts"]);
  cluster.grantDataApiAccess(fns["admin/countPagesAccounts"]);
  cluster.grantDataApiAccess(fns["admin/manageAccountLinkedUsers"]);
  cluster.grantDataApiAccess(fns["admin/listAccountLinkedUsers"]);
  cluster.grantDataApiAccess(fns["admin/countPagesUsers"]);
  return cluster;
};
