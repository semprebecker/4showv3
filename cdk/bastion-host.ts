import { Stack } from 'aws-cdk-lib';
import { BastionHostLinux, IVpc  } from 'aws-cdk-lib/aws-ec2';

import { envs } from '../backend/main/config';

export const getBastionHost = (scope: Stack, vpc: IVpc): BastionHostLinux => {
  return new BastionHostLinux(scope, `Bastion4show${envs.stage}`, { vpc: vpc, instanceName: `bastion-host-4show-${envs.stage}`})
};
