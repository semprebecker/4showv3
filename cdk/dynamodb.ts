import { RemovalPolicy, Stack } from 'aws-cdk-lib';
import { AttributeType, BillingMode, Table } from 'aws-cdk-lib/aws-dynamodb';

import { lambdaFunctions } from './lambda';
import { envs } from '../backend/main/config';

export const getTable = (stack: Stack, fns: lambdaFunctions): Table => {
  const table = new Table(stack, `Stack4showTable${envs.stage.toUpperCase()}`, {
    billingMode: BillingMode.PAY_PER_REQUEST,
    partitionKey: { name: 'pk', type: AttributeType.STRING },
    removalPolicy: RemovalPolicy.DESTROY,
    sortKey: { name: 'sk', type: AttributeType.STRING },
    tableName: `Stack4showTable${envs.stage.toUpperCase()}`,
    timeToLiveAttribute: 'exp',
  });
  table.grantReadWriteData(fns['setup-connection/getState']);
  table.grantWriteData(fns['setup-connection/onConnect']);
  table.grantWriteData(fns['setup-connection/onDisconnect']);
  return table;
};

