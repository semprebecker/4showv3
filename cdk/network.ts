import { Stack} from 'aws-cdk-lib';
import { Vpc } from 'aws-cdk-lib/aws-ec2';

import { envs } from '../backend/main/config';

export const getVpc = (scope: Stack): Vpc => {
  return new Vpc(scope, `Stack4showVPC${envs.stage}`, { natGateways: 1})
};
