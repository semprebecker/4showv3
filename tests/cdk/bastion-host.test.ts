import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { getBastionHost, } from '../../cdk/bastion-host';
import { getVpc } from '../../cdk/network';

describe('Bastion Host', () => {
  test('returns bastion-host', () => {
    const app = new App();
    const stack = new Stack(app, 'ApiTestStack', { env: { account: '123456789', region: 'us-east-1' } });
    getBastionHost(stack, getVpc(stack));
    const cfn = Template.fromStack(stack).toJSON();
    const resources = cfn.Resources;
    const matchObject: { Parameters: Record<string, unknown>; Resources: Record<string, unknown> } = {
      Parameters: expect.any(Object),
      Resources: {},
    };
    Object.keys(resources).forEach((res) => {
      switch (resources[res].Type) {
        case 'AWS::EC2::Instance':
          matchObject.Resources[res] = {
            Properties: expect.any(Object)
          };
          break;
        default:
          break;
      }
    });

    expect(cfn).toMatchSnapshot(matchObject);
  });
});
