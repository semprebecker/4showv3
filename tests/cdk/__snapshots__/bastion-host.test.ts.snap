// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bastion Host returns bastion-host 1`] = `
Object {
  "Outputs": Object {
    "Bastion4showdevBastionHostId576F7FE4": Object {
      "Description": "Instance ID of the bastion host. Use this to connect via SSM Session Manager",
      "Value": Object {
        "Ref": "Bastion4showdevFCCB4240",
      },
    },
  },
  "Parameters": Any<Object>,
  "Resources": Object {
    "Bastion4showdevFCCB4240": Object {
      "DependsOn": Array [
        "Bastion4showdevInstanceRoleDefaultPolicy1C167C0F",
        "Bastion4showdevInstanceRole4432A549",
      ],
      "Properties": Any<Object>,
      "Type": "AWS::EC2::Instance",
    },
    "Bastion4showdevInstanceProfile0697EEF5": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "Bastion4showdevInstanceRole4432A549",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "Bastion4showdevInstanceRole4432A549": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "bastion-host-4show-dev",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Bastion4showdevInstanceRoleDefaultPolicy1C167C0F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ssmmessages:*",
                "ssm:UpdateInstanceInformation",
                "ec2messages:*",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "Bastion4showdevInstanceRoleDefaultPolicy1C167C0F",
        "Roles": Array [
          Object {
            "Ref": "Bastion4showdevInstanceRole4432A549",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Bastion4showdevInstanceSecurityGroupA6EFB002": Object {
      "Properties": Object {
        "GroupDescription": "ApiTestStack/Bastion4showdev/Resource/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "bastion-host-4show-dev",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "Stack4showVPCdevB3350D88": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "Stack4showVPCdevIGW921FDFC2": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "Stack4showVPCdevPrivateSubnet1DefaultRoute703B12D1": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet1NATGateway7B76FC98",
        },
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPrivateSubnet1RouteTable338B39E6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "Stack4showVPCdevPrivateSubnet1RouteTable338B39E6": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "Stack4showVPCdevPrivateSubnet1RouteTableAssociationD947205D": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPrivateSubnet1RouteTable338B39E6",
        },
        "SubnetId": Object {
          "Ref": "Stack4showVPCdevPrivateSubnet1Subnet7DBF50C2",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "Stack4showVPCdevPrivateSubnet1Subnet7DBF50C2": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.96.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "Stack4showVPCdevPrivateSubnet2DefaultRouteB3E34412": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet1NATGateway7B76FC98",
        },
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPrivateSubnet2RouteTableBFEE83D2",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "Stack4showVPCdevPrivateSubnet2RouteTableAssociation48E19415": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPrivateSubnet2RouteTableBFEE83D2",
        },
        "SubnetId": Object {
          "Ref": "Stack4showVPCdevPrivateSubnet2SubnetE83F9285",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "Stack4showVPCdevPrivateSubnet2RouteTableBFEE83D2": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "Stack4showVPCdevPrivateSubnet2SubnetE83F9285": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.128.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "Stack4showVPCdevPrivateSubnet3DefaultRoute3515D1D6": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet1NATGateway7B76FC98",
        },
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPrivateSubnet3RouteTable63A4AAEB",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "Stack4showVPCdevPrivateSubnet3RouteTable63A4AAEB": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PrivateSubnet3",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "Stack4showVPCdevPrivateSubnet3RouteTableAssociation7397CD90": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPrivateSubnet3RouteTable63A4AAEB",
        },
        "SubnetId": Object {
          "Ref": "Stack4showVPCdevPrivateSubnet3Subnet95E9203E",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "Stack4showVPCdevPrivateSubnet3Subnet95E9203E": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1c",
        "CidrBlock": "10.0.160.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PrivateSubnet3",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "Stack4showVPCdevPublicSubnet1DefaultRoute6873F953": Object {
      "DependsOn": Array [
        "Stack4showVPCdevVPCGW61557925",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "Stack4showVPCdevIGW921FDFC2",
        },
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet1RouteTable6CDFA71C",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "Stack4showVPCdevPublicSubnet1EIPC6092CA1": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "Stack4showVPCdevPublicSubnet1NATGateway7B76FC98": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "Stack4showVPCdevPublicSubnet1EIPC6092CA1",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet1Subnet16805C75",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "Stack4showVPCdevPublicSubnet1RouteTable6CDFA71C": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "Stack4showVPCdevPublicSubnet1RouteTableAssociationE3048979": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet1RouteTable6CDFA71C",
        },
        "SubnetId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet1Subnet16805C75",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "Stack4showVPCdevPublicSubnet1Subnet16805C75": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.0.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "Stack4showVPCdevPublicSubnet2DefaultRoute6510D5DD": Object {
      "DependsOn": Array [
        "Stack4showVPCdevVPCGW61557925",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "Stack4showVPCdevIGW921FDFC2",
        },
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet2RouteTable47DFD4E7",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "Stack4showVPCdevPublicSubnet2RouteTable47DFD4E7": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "Stack4showVPCdevPublicSubnet2RouteTableAssociationEB7227BF": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet2RouteTable47DFD4E7",
        },
        "SubnetId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet2SubnetF1D03DFA",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "Stack4showVPCdevPublicSubnet2SubnetF1D03DFA": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.32.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "Stack4showVPCdevPublicSubnet3DefaultRouteF63FC35F": Object {
      "DependsOn": Array [
        "Stack4showVPCdevVPCGW61557925",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "Stack4showVPCdevIGW921FDFC2",
        },
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet3RouteTableEBB3D228",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "Stack4showVPCdevPublicSubnet3RouteTableAssociationF4930C52": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet3RouteTableEBB3D228",
        },
        "SubnetId": Object {
          "Ref": "Stack4showVPCdevPublicSubnet3Subnet5A14E65C",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "Stack4showVPCdevPublicSubnet3RouteTableEBB3D228": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PublicSubnet3",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "Stack4showVPCdevPublicSubnet3Subnet5A14E65C": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1c",
        "CidrBlock": "10.0.64.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "ApiTestStack/Stack4showVPCdev/PublicSubnet3",
          },
        ],
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "Stack4showVPCdevVPCGW61557925": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "Stack4showVPCdevIGW921FDFC2",
        },
        "VpcId": Object {
          "Ref": "Stack4showVPCdevB3350D88",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
